src/Microsoft.JSInterop/JSRuntimeBase.cs

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.JSInterop
+{
+    /// <summary>
+    /// Abstract base class for a JavaScript runtime.
+    /// </summary>
+    public abstract class JSRuntimeBase : IJSRuntime
+    {
+        private long _nextPendingTaskId = 1; // Start at 1 because zero signals "no response needed"
+        private readonly ConcurrentDictionary<long, object> _pendingTasks
+            = new ConcurrentDictionary<long, object>();
+
+        /// <summary>
+        /// Invokes the specified JavaScript function asynchronously.
+        /// </summary>
+        /// <typeparam name="T">The JSON-serializable return type.</typeparam>
+        /// <param name="identifier">An identifier for the function to invoke. For example, the value <code>"someScope.someFunction"</code> will invoke the function <code>window.someScope.someFunction</code>.</param>
+        /// <param name="args">JSON-serializable arguments.</param>
+        /// <returns>An instance of <typeparamref name="T"/> obtained by JSON-deserializing the return value.</returns>
+        public Task<T> InvokeAsync<T>(string identifier, params object[] args)
+        {
+            // TODO: Auto-timeout
```

Comment:
I'd suggest against this if you don't have a clear need in mind.  

There are lots of cases where you don't want to have a timeout - "I want to be notified if X happens, but it might never happen".

If you **want** a timeout, then you want it to be cooperative - in which case you'd want to send it to the JS side as an actual argument - but I don't believe promises have a concept like cancellation - so again, you're building your own protocol. 

If you want a timeout but you're not going to make it cooperative (client cancellation) then what you'd want to cancel is any continuations you have scheduled, which is pretty easy to do in .NET.

Reply:
OK, I'll leave it for now, but will have a comment there that it might need to be reviewed.

My concern is that in the server-executed case, it's preferable not to be leaking memory if the JS-side code is just badly written and doesn't notify about completion (e.g., on errors, or if it's a void function and the developer hasn't thought about what they are doing). Plus, an individual user could deliberately change the JS-side code running in their browser so it doesn't notify completion even if the developer wrote their app correctly.

We might decide that a timeout is not the best solution even still. Maybe there should be a default per-JS-runtime limit on the number of pending tasks. This would limit it separately for each connected user.